### ISSUE TITLE
Device still receives notifications after logout

### ISSUE THREAD

**Issue by EllinaKuznetcova** [smooch/smooch-ios#119]
I approached following error: when user logged in to smooch account from 2 devices and if user logged out from one of them he still receives notifications on both devices. Is there a way to fix this? I am using `Smooch.logout()` to logout user.

**Comment by mspensieri** [comment]
@EllinaKuznetcova could you please provide a bit more information so that I can find the user in question and help debug this further?

* A specific `userId` that is still receiving notifications after logout
* The email address associated to your smooch account to help me find your user
* Sample code showing how/where you are using login/logout
* The SDK version you are using

Thanks!

**Comment by EllinaKuznetcova** [comment]
@mspensieri Here the information you needed:

- `userId`: 222
-  `email`: ellina@t.t
- sdk version: 5.1.0
- sample code

for login (happens before conversation controller creating and when user logged in to app):
```
        let settings = SKTSettings(appToken: smoochToken)
        settings.userId = "\(user.id)"
        Smooch.initWith(settings)
        Smooch.login("\(user.id)", jwt: nil)
        SKTUser.current()?.firstName = user.firstName
        SKTUser.current()?.lastName = "\(user.id)"
        SKTUser.current()?.email = user.email
```

for logout (happens when server token expired or when user log out. redirect to main app screen):
```
static func logout() {
        guard let user = ENUser.getMainUser() else {return}
        user.token = nil
        do {
            let realm = try Realm()
            try realm.write {
                realm.delete(realm.objects(ENUser.self))
            }
            Smooch.logout()
        }
        catch let error {
            Logger.error(error)
        }
    }
```

Thanks!

**Comment by mspensieri** [comment]
Hmm I wasn't able to find any account under `ellina@t.t`, is that email correct?

**Comment by EllinaKuznetcova** [comment]
I am sure it is correct, is finding user by id or name will be an option? 
User ID: 222
Name: Ellina 222
Email: ellina@t.t

**Comment by mspensieri** [comment]
I mean the email of your account (the one you use to log in to app.smooch.io), not the email of the user on your app. User Ids are not unique across Smooch app tokens, so in order to find your specific user I'll need to first identify which account I'm looking at

**Comment by EllinaKuznetcova** [comment]
Oh, I am sorry. This one should work: kevin@b-designworks.com .

**Comment by mspensieri** [comment]
Ok thanks. I was able to find your user, and it seems that there are multiple push notification tokens configured for what looks to be the same physical device.

The user in question has 5 devices on record, 3x `iPhone6,2` (iOS `10.1.1`), and 2x `iPhone7,1` (iOS `10.0.2`). Im assuming those 5 actually map to two real, physical devices since the specs match exactly for the multiple occurrences. Can you confirm that you have only two devices for testing?

Of those 5 "devices", 4 of them have unique push tokens configured, 2 of which are still active. One of the "devices" has no push token configured. So what I believe happened is that your user has multiple push tokens configured for the same physical device, but logging out only deactivates one of those, leaving the other one as still active.

So, how can this happen?

Smooch stores the unique device identifier in `NSUserDefaults`, and that uniquely identifies a device (i.e. a push token) across users. If `NSUserDefaults` gets wiped for some reason, Smooch will create a new unique device identifier and consider it as an entirely different device. Unfortunately, there's a known issue on iOS 9.3.1 and later that can periodically cause `NSUserDefaults` to appear empty, which will trigger the aforementioned behaviour and leave your user in an inconsistent state, which manifests itself as the push bug you're seeing.

https://forums.developer.apple.com/thread/44264
https://forums.developer.apple.com/message/135644
https://forums.developer.apple.com/message/135342

In this case I can clean up the user for you if you want. I'll also bring this issue to the team to further discuss what can be done to mitigate this issue in the future

**Comment by EllinaKuznetcova** [comment]
Unfortunately, I cannot give you a number of devices in use, because part of them on customer side. 

As for cleaning user, am I right: when I create new smooch user issue with tokens will still be exist? In that case cleaning up user will be an option. 

Thank you!

**Comment by mspensieri** [comment]
I'm not sure I fully understand your question. What I can do is remove the active push notification tokens from that user, so that it fixes the issue for now. Next time you log in after the cleanup, push notifications will be re-enabled on that device

**Comment by mspensieri** [comment]
A fix for this issue was made in [v5.4.0](https://github.com/smooch/smooch-ios/releases/tag/5.4.0). It won't fix any currently affected users, but it should prevent this from happening again for new users
